File: bridge/python/src/destack/_impl/mir/metadata/drop.py
from __future__ import annotations

from destack._impl.model import ModelImpl


class DropMetadataImpl(ModelImpl):
    """Methods for drop metadata."""

    def drop_glue(self, ty):
        """Return full drop glue for a type."""

        return self.glue_by_type.get(ty)

    def drop_hook(self, ty):
        """Return the user-authored drop hook for a type."""

        return self.hooks_by_type.get(ty)


class DropGlueImpl(ModelImpl):
    """Methods for drop glue."""

    def function_id(self):
        """Return the concrete function backing this glue."""

        if self.kind == "generated":
            return self.function

        return None

    def is_generated_function(self, target) -> bool:
        """Return whether this glue is generated for the given function."""

        if self.kind != "generated":
            return False

        return self.function == target


class DropHookImpl(ModelImpl):
    """Methods for drop hooks."""

    def is_function(self, target) -> bool:
        """Return whether this hook calls the given function."""

        return self.function == target

File: client/typescript/src/_impl/mir/metadata/drop.ts
import type {
    DropGlue,
    DropHook,
    DropMetadata,
} from "../../../_generated/mir/metadata/drop.js";
import type { LocalNodeId } from "../../../_generated/mir/tree/node.js";
import { getLocalNodeValue, localNodeIdMatches } from "../node.js";

export const DropMetadataImpl = {
    /** Return full drop glue for a type. */
    dropGlue(metadata: DropMetadata, ty: LocalNodeId): DropGlue | undefined {
        return getLocalNodeValue(metadata.glueByType, ty);
    },

    /** Return the user-authored drop hook for a type. */
    dropHook(metadata: DropMetadata, ty: LocalNodeId): DropHook | undefined {
        return getLocalNodeValue(metadata.hooksByType, ty);
    },
};

export const DropGlueImpl = {
    /** Return the concrete function backing this glue. */
    functionId(glue: DropGlue): LocalNodeId | undefined {
        return glue.kind === "generated" ? glue.function : undefined;
    },

    /** Return whether this glue is generated for the given function. */
    isGeneratedFunction(glue: DropGlue, target: LocalNodeId): boolean {
        return glue.kind === "generated" && localNodeIdMatches(glue.function, target);
    },
};

export const DropHookImpl = {
    /** Return whether this hook calls the given function. */
    isFunction(hook: DropHook, target: LocalNodeId): boolean {
        return localNodeIdMatches(hook.function, target);
    },
};
